File: services/ai/src/components/game/game.ts
import { Rank, Tensor, TensorBuffer } from "@tensorflow/tfjs-node";
import * as Tf from "@tensorflow/tfjs-node";
import { e_GAME_CONSTANTS, e_TAG_PLAYER, e_TYPE_MESSAGE, PARAMS } from "@constants/constants";
import { BallState, e_GAME_CHANNEL, e_GAME_STATE, PlayerState, StateNotificationMsg, StateResponseMsg } from "@ai-types/t_game";
import { MessageGame } from "@ai-types/t_game";
import { Observer } from "@segfaultx/observable";

export class	PongGame {

	private	_url: string = "";
	private	last_message: void | Promise<void> = undefined;
	private	listener: Observer<boolean> | undefined;

	get	hsquares() {return (30)}
	get	vsquares() {return (30)}
	get	channels() {return (3)}
	get fields() {return (8)}
	set	url(url: string) {
		/*[PENDING]: Set properly the regex.*/
		if (url.length == 0)//if (/https?:\/\/.*/.test(url))
			throw (new Error("Must be defined an url"));
		this._url = url;
	}

	constructor (url: string) {
		this.url = url;
	}

	private	ft_setAbsolutePositions(position: {pos_x?: number, pos_y?: number}[]): void {
		position.forEach(position => {
			if (position.pos_x != undefined)
				position.pos_x = e_GAME_CONSTANTS.WIDTH / 2 + position.pos_x;
			if (position.pos_y != undefined)
				position.pos_y = e_GAME_CONSTANTS.HEIGHT / 2 - position.pos_y;
		});
	}

	public	ft_reset(): void {
		this.last_message = undefined;
	}

	private	ft_fillPlayerOnDiscreteState(buf_racquet: TensorBuffer<Rank, "float32">,
										 player: PlayerState, index: number): void {
		const	hratio = this.vsquares / e_GAME_CONSTANTS.HEIGHT;
		const	center = Math.floor(player.pos_y * hratio);
		const	half_racquet = Math.round((e_GAME_CONSTANTS.HALF_RACQUET * hratio));
		let		rq_start = center - half_racquet;
		rq_start = rq_start < 0 ? 0 : rq_start;
		let		rq_end = center + half_racquet;
		rq_end = rq_end > (this.vsquares - 1) ? this.vsquares : rq_end;
		for (;rq_start <= rq_end; rq_start++) {
			switch (player.tag) {
				case e_TAG_PLAYER.ONE: 
					buf_racquet.set(1, index, rq_start, 0, e_GAME_CHANNEL.RONE_CH);
					break ;
				case e_TAG_PLAYER.TWO:
					const	pos_x = this.hsquares - 1;
					buf_racquet.set(1, index, rq_start, pos_x, e_GAME_CHANNEL.RTWO_CH);
					break ;
				default:
					return ;
			}
		}
	}

	/*[PENDING]: Test if this implementation to Array is well implemented.*/
	public	ft_getStateTensor(state: StateResponseMsg[]): Tensor {
		const	buffer = Tf.buffer([state.length, this.vsquares, this.hsquares, this.channels]);
		state.forEach((state, index) => {
			const	state_cloned = JSON.parse(JSON.stringify(state));
			const	ball: BallState = state_cloned.body.ball; 
			const	players: PlayerState[] = state_cloned.body.players;
			const	hratio = (this.hsquares - 1) / e_GAME_CONSTANTS.WIDTH;
			const	vratio = (this.vsquares - 1) / e_GAME_CONSTANTS.HEIGHT;
			this.ft_setAbsolutePositions([ball]);
			this.ft_setAbsolutePositions(players);
			const	cx_square = Math.floor(ball.pos_x * hratio);
			const	cy_square = Math.floor(ball.pos_y * vratio);
			buffer.set(1, index, cy_square, cx_square, e_GAME_CHANNEL.BALL_CH);
			this.ft_fillPlayerOnDiscreteState(buffer, players[e_TAG_PLAYER.ONE]!, index);
			this.ft_fillPlayerOnDiscreteState(buffer, players[e_TAG_PLAYER.TWO]!, index);
		});
		return (buffer.toTensor());
	}

    private	ft_normalizeState(state: StateResponseMsg): StateResponseMsg {
        const	st = JSON.parse(JSON.stringify(state));
        const	vel_x = st.body.ball.vel_x;
        const	vel_y = st.body.ball.vel_y;
        let		vel_mg = Math.sqrt(vel_x * vel_x + vel_y * vel_y)
        if (vel_mg == 0)
            vel_mg = 1.0
        st.body.ball.pos_x /= (e_GAME_CONSTANTS.WIDTH / 2)
        st.body.ball.pos_y /= (e_GAME_CONSTANTS.HEIGHT / 2)
        st.body.players.forEach((player: PlayerState) => {
            player.pos_y /= (e_GAME_CONSTANTS.HEIGHT / 2)
		});
        st.body.ball.vel_x /= vel_mg
        st.body.ball.vel_y /= vel_mg
        return (st);
	}

    public	ft_getStateTensorMLP(states: StateResponseMsg[]): Tf.Tensor {

		const	buffer = Tf.buffer([states.length, this.fields]);
		states.forEach((st, i) => {
			if (st.body.players.length == 0 || st.body.players == undefined)
				return (buffer.toTensor());
			const	ball = st.body.ball;
			const	players = st.body.players;
            //The sign is significant because give us information about where is the ball and myself.
            const	discriminant_y = (ball.pos_y - players[e_TAG_PLAYER.TWO]!.pos_y) / e_GAME_CONSTANTS.HEIGHT
            //The sign here is not significant, because we just need to know how far is the ball from us.
            const	discriminant_x = Math.abs(ball.pos_x - PARAMS["rigth_bound"]) / e_GAME_CONSTANTS.WIDTH
            const	st_norm = this.ft_normalizeState(st)
            const	ball_nm = st_norm.body.ball;
            const	players_nm = st_norm.body.players;
			//Set the vector normalized in the buffer.
            buffer.set(ball_nm.pos_x, i, 0);
			buffer.set(ball_nm.pos_y, i, 1);
			buffer.set(ball_nm.vel_x, i, 2);
			buffer.set(ball_nm.vel_y, i, 3);
			buffer.set(players_nm[0]!.pos_y, i, 4);
			buffer.set(players_nm[1]!.pos_y, i, 5);
			buffer.set(discriminant_y, i, 6);
			buffer.set(discriminant_x, i, 7);
		});
        return (buffer.toTensor()); 
	}

	public	ft_subscribeToSocket(ob: Observer<MessageGame>) : WebSocket | undefined {
		/*[!PENDING]: How to handle errors from initialization.*/
		/*[!PENDING]: How to handle the other events*/
		try {
			const	socket = new WebSocket(this._url);
			socket.addEventListener("message", (ev: MessageEvent<any>) => {
				/*[!PENDING][!IMPORTANT]: What to do if the client can't read
				* messages from a network connection problem? I think the connection
				* is closed by the socket server and that will notify the other client
				* and also the server will stop.*/
				if (this.last_message === undefined) {
					this.last_message = ob.ft_update(JSON.parse(ev.data));
				}
				else {
					if (this.last_message instanceof Promise) {
						this.last_message = this.last_message.then(async () => {
							const	msg: MessageGame = JSON.parse(ev.data);
							await ob.ft_update(msg);
							if (msg.type == e_TYPE_MESSAGE.NOTIFICATION) {
								const	noti = msg as StateNotificationMsg;
								if (noti.body.status == e_GAME_STATE.FINISH)
									await this.listener?.ft_update(true);
							}
						});
					}
				}
			});
			return (socket);
		}
		catch (err: unknown) {
		}
		return (undefined);
	}

	public	ft_listenWhenDone(ob: Observer<boolean>): void {
		this.listener = ob;
	}
}

File: services/ai_python/src_py/components/game/game.py
import json
import threading
from typing import List, Optional
import numpy as np
import tensorflow as tf
from websocket import WebSocketApp

from src_py.constants.constants import e_GAME_CONSTANTS, e_TAG_PLAYER, e_TYPE_MESSAGE, e_GAME_STATE
from src_py.types.t_game import (
    BallState, PlayerState, StateResponseMsg, MessageGame, StateNotificationMsg
)
from src_py.utils.observable import Observer

class PongGame:
    def __init__(self, url: str):
        self._url: str = ""
        self.url = url
        self._last_message_lock = threading.Lock()
        self._listener: Optional[Observer[bool]] = None
        self._ws: Optional[WebSocketApp] = None
        self._ws_thread: Optional[threading.Thread] = None

    @property
    def hsquares(self) -> int: return 30
    @property
    def vsquares(self) -> int: return 30
    @property
    def channels(self) -> int: return 3

    @property
    def url(self) -> str: return self._url
    @url.setter
    def url(self, url: str):
        if not url:
            raise ValueError("Must be defined an url")
        self._url = url

    def ft_reset(self) -> None:
        # nothing persistent besides the socket in this client
        pass

    def _ft_setAbsolutePositions(self, positions: List[dict]) -> None:
        for p in positions:
            if "pos_x" in p and p["pos_x"] is not None:
                p["pos_x"] = e_GAME_CONSTANTS.WIDTH / 2 + p["pos_x"]
            if "pos_y" in p and p["pos_y"] is not None:
                p["pos_y"] = e_GAME_CONSTANTS.HEIGHT / 2 - p["pos_y"]

    def _ft_fillPlayerOnDiscreteState(self, buf: np.ndarray, player: PlayerState, batch_index: int) -> None:
        hratio = self.vsquares / e_GAME_CONSTANTS.HEIGHT
        center = int(np.floor(player.pos_y * hratio))
        half_racquet = int(np.round(e_GAME_CONSTANTS.HALF_RACQUET * hratio))
        rq_start = max(0, center - half_racquet)
        rq_end = min(self.vsquares - 1, center + half_racquet)
        for y in range(rq_start, rq_end + 1):
            if player.tag == e_TAG_PLAYER.ONE:
                buf[batch_index, y, 0, 1] = 1.0
            elif player.tag == e_TAG_PLAYER.TWO:
                x = self.hsquares - 1
                buf[batch_index, y, x, 2] = 1.0

    def ft_getStateTensor(self, states: List[StateResponseMsg]) -> tf.Tensor:
        # buffer shape: [batch, H, W, C]
        buffer = np.zeros((len(states), self.vsquares, self.hsquares, self.channels), dtype=np.float32)
        for i, st in enumerate(states):
            # deep copy via json; we won't mutate original
            st_copy = json.loads(json.dumps(st, default=lambda o: o.__dict__))
            ball = st_copy["body"]["ball"]
            players = st_copy["body"]["players"]
            # to absolute
            self._ft_setAbsolutePositions([ball])
            self._ft_setAbsolutePositions(players)

            hratio = (self.hsquares - 1) / e_GAME_CONSTANTS.WIDTH
            vratio = (self.vsquares - 1) / e_GAME_CONSTANTS.HEIGHT
            cx = int(np.floor(ball["pos_x"] * hratio))
            cy = int(np.floor(ball["pos_y"] * vratio))
            buffer[i, cy, cx, 0] = 1.0

            # players back to dataclass-like access
            for p in players:
                ps = PlayerState(
                    pos_y=float(p["pos_y"]),
                    action=p["action"],
                    tag=p["tag"],
                    score=int(p["score"])
                )
                self._ft_fillPlayerOnDiscreteState(buffer, ps, i)

        return tf.convert_to_tensor(buffer)

    def ft_subscribeToSocket(self, ob: Observer[MessageGame]):
        def _on_message(ws, message):
            msg = json.loads(message)
            # forward to observer
            ob.ft_update(msg)
            # notify done -> trainer listener
            if msg.get("type") == e_TYPE_MESSAGE.NOTIFICATION:
                noti = msg
                if noti.get("body", {}).get("status") == e_GAME_STATE.FINISH and self._listener:
                    self._listener.ft_update(True)

        def _on_error(ws, err):
            # Optional: log or escalate
            pass

        def _on_close(ws, code, reason):
            pass

        def _on_open(ws):
            pass

        self._ws = WebSocketApp(
            self._url,
            on_open=_on_open,
            on_message=_on_message,
            on_error=_on_error,
            on_close=_on_close
        )
        self._ws_thread = threading.Thread(target=self._ws.run_forever, daemon=True)
        self._ws_thread.start()
        return self._ws  # parity with TS: return socket-like handle

    def ft_listenWhenDone(self, ob: Observer[bool]) -> None:
        self._listener = ob
